collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java

```diff
@@ -0,0 +1,352 @@
+                        + "to login to your network?";
+
+                throw new Exception(error);
+            }
+        }
+
+        InputStream downloadStream = body.byteStream();
+
+        String hash = "";
+
+        if (HTTP_CONTENT_TYPE_TEXT_XML.equals(contentType)) {
+            byte[] bytes = IOUtils.toByteArray(downloadStream);
+            downloadStream = new ByteArrayInputStream(bytes);
+            hash = FileUtils.getMd5Hash(new ByteArrayInputStream(bytes));
+        }
+
+        Map<String, String> responseHeaders = new HashMap<>();
+        Headers headers = response.headers();
+
+        if (headers.size() > 0) {
+            for (int i = 0; i < headers.size(); i++) {
+                responseHeaders.put(headers.name(i), headers.value(i));
+            }
+        }
+
+        return new HttpGetResult(downloadStream, responseHeaders, hash, statusCode);
+    }
+
+    @NonNull
+    @Override
+    public HttpHeadResult head(@NonNull URI uri, @Nullable HttpCredentialsInterface credentials) throws Exception {
+        Map<String, String> responseHeaders = new HashMap<>();
+
+        int statusCode;
+
+        try {
+            Timber.i("Issuing HEAD request to: %s", uri.toString());
+
+            OkHttpClient client = createOkHttpClient(credentials);
+            Request request = headRequest(uri);
+
+            Response response = client.newCall(request).execute();
+            statusCode = response.code();
+
+            if (statusCode == HttpURLConnection.HTTP_NO_CONTENT) {
+                Headers headers = response.headers();
+
+                for (String headerName : headers.names()) {
+                    responseHeaders.put(headerName, headers.get(headerName));
+                }
+            }
+
+            discardEntityBytes(response);
+
+        } catch (IOException | IllegalStateException e) {
+            String errorMessage = "";
+
+            if (e instanceof MalformedURLException) {
+                errorMessage = "Malformed URL Exception";
+            } else if (e instanceof IllegalStateException) {
+                errorMessage = "Illegal State Exception";
+            }
+
+            throw new Exception(errorMessage);
+        } catch (Exception e) {
+            String msg = e.getMessage();
+            if (msg == null) {
+                msg = e.toString();
+            }
+
+            throw new Exception("Generic Exception: " + msg);
+        }
+
+        return new HttpHeadResult(statusCode, responseHeaders);
+    }
+
+    @NonNull
+    @Override
+    public ResponseMessageParser uploadSubmissionFile(@NonNull List<File> fileList, @NonNull File submissionFile, @NonNull URI uri, @Nullable HttpCredentialsInterface credentials) throws IOException {
+        ResponseMessageParser messageParser = null;
+
+        boolean first = true;
+        int fileIndex = 0;
+        int lastFileIndex;
+        while (fileIndex < fileList.size() || first) {
```

Comment:
This whole section for building the parts could use some refactoring. Not urgent and a good start might be to add a comment or two. Things like, on the `first` flag, "Always add the submission XML, even if there are no attachments" and on the `while` loop, "split the files to send into parts below the part length threshold"

Reply:
Agreed, I refactored it a little bit - but was more concerned with keeping the logic as close as possible. Could this be done under a new Issue number / PR?